service/paging.go

```diff
@@ -1,55 +1,74 @@
 package service
 
 import (
-	"fmt"
 	"net/http"
-	"strconv"
+	"strings"
 )
 
+type Sort struct {
+	Field     string
+	Direction string
+}
+
+func (s Sort) Valid(fieldWhiteList []string) bool {
+	if s.Field == "" || !contains(fieldWhiteList, s.Field) {
+		return false
+	}
+
+	switch s.Direction {
+	case "asc",
+		"desc":
+	// no-op
+	default:
+		return false
+	}
+
+	return true
+}
+
 // PagingParams represents paging and sorting parameter values
 type PagingParams struct {
 	PageSize   *int32
-	PageNumber *int32
-	SortField  *string
-	SortAsc    *bool
+	Offset     *int32
+	SortFields []Sort
 }
 
 // ParsePagingParams retrieves paging params from the request, allows for whitelisting sort field values
-func ParsePagingParams(r *http.Request, sortFieldsWhitelist []string) (PagingParams, error) {
-	paging := PagingParams{}
+func ParsePagingParams(r *http.Request, defaults PagingParams, sortFieldsWhitelist []string) (PagingParams, error) {
+	paging := defaults
 
-	// page number
-	pageNumber, err := Int32PointerFromQueryParam(r, "page_number")
+	// offset (page number)
+	offset, err := Int32PointerFromQueryParam(r, "offset")
 	if err != nil {
 		return paging, err
 	}
-	paging.PageNumber = pageNumber
+	if offset != nil {
```

Comment:
So I think this is where my confusion was yesterday when we were talking. If we are going to have default behavior here then how do we know if they wanted default or if they wanted to "get all". I don't know how I feel about default behavior being "get all" because that is how I think we would have to treat it.

Reply:
Been turning this over in my head. I don't think we want this parser to be concerned with "get all" or what not. It should just parse things out and give you back what it finds -- either the empty value, the default you passed in, or the value from the actual query string.

The service should decide how it is going to behave. If you don't pass in offset and page size but wants pagination, it should pass in defaults for those values to ensure you are always using pagination. If it wants to do get all, it doesn't pass in defaults and then assumes empty values coming back from the parse indicates the caller is fine with "get all".

Does that make sense?